### ISSUE TITLE
Make HttpClient's authenticate and authenticateProxy callbacks accept `null` realm.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/sdk#44039]
The migration of `dart:io` made the callbacks set by `HttpClient.authenticate` and `HttpClient.authenticateProxy` accept a non-nullable `realm` argument. However, the `realm` can be `null`, so this was a mistake.
The mistake was hidden by the implementation in `_HttpClient` storing the functions as `Function`-typed fields, and therefore not recognizing that they were being called with a nullable `realm` string.
See https://github.com/dart-lang/sdk/issues/42870 for a bug from this behavior.

The change is to make the methods have signatures:
```dart
  void set authenticateProxy(Future<bool> f(String host, int port, String scheme, String? realm)?);
  void set authenticate(Future<bool> f(Uri url, String scheme, String? realm)?);
```
This is a breaking change. If existing code calls the setter with a function not accepting a nullable `realm`, it will be a compile-time error. 
(If the function already accepts a nullable `realm`, it will keep working. If the function is a function literal with no type argument on the `realm`, it will change to being nullable, which may or may not cause the function body to be an error).

The function *is* already called with a `null` occasionally, as the bug report shows. That is why we need to change the signatures to match the actual behavior.
This also means that the only code which will be broken is code where the callback function does not accept a nullable `realm`, and where the usage somehow ensures that it's not called with `null` either.

The impact is expected to be very low. This is not widely used functionality, and it only affects code which has already been migrated to null safety. It's quite possible that there is no null-safe code using the functionality outside of our own tests.
A search of Google internal Dart code shows only one usage of `.authenticate =`, and there is one usage in `package:http_io` package's `test` directory. Neither has been migrated yet, so they are unaffected by the change.

Mitigation means adding `?` to the type of `realm` in callbacks, or removing the type where possible, and only affects already null safe code.

**Comment by lrhn** [comment]
Proposed fix: https://dart-review.googlesource.com/c/sdk/+/170094

**Comment by franklinyow** [comment]
CC:   @mit-mit @vsmenon @kevmoo @Hixie @matanlurey

**Comment by mit-mit** [comment]
SGTM

**Comment by franklinyow** [comment]
Ping @vsmenon @matanlurey @Hixie  for approval

**Comment by vsmenon** [comment]
lgtm

**Comment by Hixie** [comment]
I have no opinion.

**Comment by franklinyow** [comment]
Approved

**Comment by franklinyow** [comment]
@lrhn What's the status of this change?

**Comment by itsjustkevin** [comment]
Checking in on this as there has been no movement since May.  @Hixie @vsmenon @grouma

**Comment by grouma** [comment]
LGTM.

**Comment by Hixie** [comment]
I continue to have no opinion. :-)

**Comment by vsmenon** [comment]
@lrhn - this one is approved - is it still valid?  :-)

**Comment by itsjustkevin** [comment]
@lrhn do we still want to move forward with this breaking change?

**Comment by vsmenon** [comment]
@brianquinlan - is this on your radar?

**Comment by itsjustkevin** [comment]
Friendly ping @lrhn and @brianquinlan.  I am doing a cleanup and want to make sure this is still valid.

**Comment by brianquinlan** [comment]
It was not on my radar but it was fixed in cf627eb01ce08eb7cea2240e206b8710ce4bb783
